src/Faithlife.Testing/AssertEx.cs

```diff
@@ -842,6 +836,69 @@ private static string ToString(object obj)
 			return toString;
 		}
 
+		private static string ToPrettyJson(object obj, int tabLevel)
+		{
+			var rawJson = JsonUtility.ToJson(obj, s_jsonWithTabs).Replace("\r\n", "\n");
+			var lines = rawJson.Split('\n').Select(l => l.Trim()).ToList();
+
+			// if json can fit on one line, return inlined
+			var shortJson = lines.Select(l => l + (l.EndsWithOrdinal(",") ? " " : "")).Join("");
+			if (shortJson.Length < c_maxJsonLength)
+				return shortJson;
+
+			// else construct pretty json output
+			var longJson = "";
+			for (var i = 0; i < lines.Count; i++)
+			{
+				var line = lines[i];
+				var hasArrayStart = line.EndsWithOrdinal("[");
+				var hasStart = hasArrayStart || line.EndsWithOrdinal("{");
+				var hasEnd = line.StartsWithOrdinal("]") || line.StartsWithOrdinal("}");
+
+				if (hasArrayStart)
+				{
+					// try to inline the array if it's short and composed of primitives
+					var arrayLines = new List<string>();
+					var hasNested = false;
+					var j = i;
+					while (++j < lines.Count)
+					{
+						var arrayLine = lines[j];
+						if (arrayLine.StartsWithOrdinal("]"))
+							break;
+
+						if (arrayLine.StartsWithOrdinal("[") || arrayLine.StartsWithOrdinal("{"))
+						{
+							hasNested = true;
+							break;
+						}
+
+						arrayLines.Add(arrayLine);
+					}
+
+					var shortArray = line + arrayLines.Join(" ") + lines[j];
```

Comment:
When the above `while (++j < lines.Count)` loop terminates because `j == lines.Count`, this `lines[j]` will throw.

If you want to communicate that you expect the loop to always terminate because well-formed JSON implies we'll always hit one of the `break` conditions, I'd recommend a `while (true)` loop and/or an explicit exception referencing what unexpected JSON formation caused the problem.